### PR TITLE
feat: Variable-Beam-Width-Search (VBWS) Part2

### PR DESCRIPTION
**Pull request by wili-65535**
## Part 2 of VBWS support.
+ Part 1 is here: [PR3082](https://github.com/NVIDIA/TensorRT-LLM/pull/3082).

## Target of this PR

+ Add infrastructure for Beam Search layer and kernels for VBWS.
+ Add corresponding unit tests for Beam Search layer and kernels, such as `cpp/tests/unit_tests/layers/beamSearchLayerTest.cpp`.
+ Adjust some classes (such as `class DecoderDomain`) to support VBWS (but they are not enabled in normal Beam Search workflow now).

### CODE REVIEW

File: cpp/tensorrt_llm/kernels/beamSearchKernels.h [hunk 1]
```diff
@@ -116,13 +128,95 @@ __device__ __forceinline__ T applyLengthPenalty(T const log_prob, int const leng
...
+#define QD2(x, nElement, nCol) QD(x, ((nElement) / (nCol)), nCol, nCol)
+
+// Use only in device function
+#define GUARD(blockIdxx, bSync, code)                                                                                  \
+    {                                                                                                                  \
+        if (bSync)                                                                                                     \
+        {                                                                                                              \
+            __syncthreads();                                                                                           \
+        }                                                                                                              \
+        if (true && blockIdx.x == (blockIdxx) && blockIdx.y == 0 && blockIdx.z == 0 && threadIdx.x == 0                \
```
Comment (byshiue) [inline cpp/tensorrt_llm/kernels/beamSearchKernels.h]: What's the motivation to add "true &&" here?

File: cpp/tensorrt_llm/layers/beamSearchLayer.cu [hunk 2]
```diff
@@ -32,38 +32,38 @@ namespace tensorrt_llm::layers
 
 #define GET_INFO_STAGE1(nPBM)                                                                                          \
     {                                                                                                                  \
-        int constexpr nBlock = (nPBM < 16) ? ((nPBM < 8) ? kThreadForSmallBeamWidth : 128) : 64;                       \
+        int constexpr nBlock = ((nPBM) < 16) ? (((nPBM) < 8) ? kThreadForSmallBeamWidth : 128) : 64;                   \
```
Comment (byshiue) [inline cpp/tensorrt_llm/layers/beamSearchLayer.cu]: What's the reason to change nPBM to (nPBM)?

File: cpp/tensorrt_llm/layers/beamSearchLayer.cu [hunk 3]
```diff
@@ -265,55 +276,24 @@ void BeamSearchLayer<T>::setup(SizeType32 const batchSize, SizeType32 const beam
...
         mBeamSearchDiversityRateHost, mBeamSearchDiversityRateDevice, batchSlots, std::make_pair(-fltEpsilon, fltMax),
         "diversity rate");
     fillBuffers(setupParams->lengthPenalty, DefaultDecodingParams::getLengthPenalty(), mLengthPenaltyHost,
         mLengthPenaltyDevice, batchSlots, std::make_pair(fltMin, fltMax), "length penalty");
     fillBuffers(setupParams->earlyStopping, DefaultDecodingParams::getEarlyStopping(), mEarlyStoppingHost,
-        mEarlyStoppingDevice, batchSlots, std::make_pair(-fltEpsilon, std::numeric_limits<int>::max()),
-        "early stopping");
+        mEarlyStoppingDevice, batchSlots, std::make_pair(-fltEpsilon, int32Max), "early stopping");
+    fillBuffers(setupParams->beamWidthArray, DefaultDecodingParams::getBeamWidthArray(), mBeamWidthArrayHost,
+        mBeamWidthArrayDevice, batchSlots, std::make_pair(-fltEpsilon, int32Max), "beam width array");
```
Comment (byshiue) [inline cpp/tensorrt_llm/layers/beamSearchLayer.cu]: Why the max value of beamWidthArray is int32Max?

File: cpp/tensorrt_llm/layers/beamSearchLayer.h [hunk 4]
```diff
@@ -56,13 +56,18 @@ class BeamSearchLayer : public BaseLayer
...
-    TensorPtr mBeamSearchDiversityRateHost;   // [batchSize], in pinned host memory.
-    TensorPtr mLengthPenaltyHost;             // [batchSize], in pinned host memory.
-    TensorPtr mEarlyStoppingHost;             // [batchSize], in pinned host memory.
+    TensorPtr mBeamSearchDiversityRateHost;   // [batchSize] cpu
+    TensorPtr mBeamSearchDiversityRateDevice; // [batchSize] gpu
+    TensorPtr mLengthPenaltyHost;             // [batchSize] cpu
+    TensorPtr mLengthPenaltyDevice;           // [batchSize] gpu
+    TensorPtr mEarlyStoppingHost;             // [batchSize] cpu
+    TensorPtr mEarlyStoppingDevice;           // [batchSize] gpu
+    TensorPtr mBeamWidthArrayHost;            // [batchSize] cpu
```
Comment (byshiue) [inline cpp/tensorrt_llm/layers/beamSearchLayer.h]: What's difference between `mBeamWidthArray` and `mBeamWidthIn`

Comment (wili-65535) [reply cpp/tensorrt_llm/layers/beamSearchLayer.h]: It's typo in comments, it should be like:
`BeamSearchLayer::mBeamWidthArrayHost` with shape [batchSize, 8]

===(gather beamWidth from each request for the current step)===>

`BeamSearchLayer::mBeamWidthIn` / `BeamSearchLayer::mBeamWidthOut` with shape [batchSize]

Assign in `beamSearchLayer.cu`:

`bh.beamWidthArraysHost = mBeamWidthArrayHost`  with shape [batchSize, 8]
`bh.beamWidthArraysDevice = mBeamWidthArrayDevice`  with shape [batchSize, 8]

`bh.nBeamWidthInHost = BeamSearchLayer::mBeamWidthIn`  with shape [batchSize]
`bh.nBeamWidthOutHost = BeamSearchLayer::mBeamWidthOut`  with shape [batchSize]

`bh.nBeamWidthInDevice` and `bh.nBeamWidthOutDevice` are for later Beam Search with diverse beam width (kernels will use these).

File: cpp/tensorrt_llm/layers/beamSearchLayer.cu (hunk 3 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/layers/beamSearchLayer.cu]: Fixed as `kMaxBeamWidth` (1024).

File: cpp/tensorrt_llm/layers/beamSearchLayer.cu (hunk 2 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/layers/beamSearchLayer.cu]: A more parenthesis for the argument in macro.

In a older commit, we gave a expression as the argument like `GET_INFO_STAGE1(x+y)` and the result was incorrect.

But as your notice, this parenthesis could be removed since we use the macros with only `nPBM` as argument in all places.

File: cpp/tensorrt_llm/kernels/beamSearchKernels.h (hunk 1 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/kernels/beamSearchKernels.h]: Typo during debug, I will remove it.

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp [hunk 5]
```diff
@@ -131,53 +205,83 @@ void BaseSamplingLayerTest<T>::setup(uint64_t seed, TestSamplingParams const& pa
...
-
-    decodeInputTensors->finished = mFinishedDevice;
-
-    decodeInputTensors->probsComputed = mComputeProbs;
-
-    decodeInputTensors->curandStates = reinterpret_cast<curandState_t*>(bufferCast<int8_t>(*mCurandStatesDevice));
-
-    return decodeInputTensors;
+    if (mBeamWidth > 1)
+    {
```
Comment (byshiue) [inline cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:216]: The same parts can be reused.

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp [hunk 6]
```diff
@@ -131,53 +205,83 @@ void BaseSamplingLayerTest<T>::setup(uint64_t seed, TestSamplingParams const& pa
...
+        decodeOutputs->beamHypotheses->numBeamsCBA = bufferCast<int>(*mNumBeamsCBA);
+        decodeOutputs->beamHypotheses->minNormedScoresCBA = bufferCast<float>(*mMinNormedScoresCBA);
+        decodeOutputs->beamHypotheses->batchDones = bufferCast<bool>(*mBatchDones);
+
+        return decodeOutputs;
+    }
+    else
+    {
+        auto decodeOutputs = std::make_shared<BaseDecodingOutputs>(mOutputIdsDevice);
+        decodeOutputs->outputIdsPtr = mIdsPtrHost;
```
Comment (byshiue) [inline cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:269]: Should it be mIdsPtrHost? Why using beam search or not are different?

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp [hunk 7]
```diff
@@ -131,53 +205,83 @@ void BaseSamplingLayerTest<T>::setup(uint64_t seed, TestSamplingParams const& pa
...
 
-    decodeOutputs->finished = mFinishedDevice;
-
-    decodeOutputs->outputLogProbs = mOutputLogProbsDevice;
-
-    decodeOutputs->cumLogProbs = mCumLogProbsDevice;
-
-    // TODO(nkorobov): check log probs and cum_log_probs
-    return decodeOutputs;
+    if (mBeamWidth > 1)
```
Comment (byshiue) [inline cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:241]: It looks the beamWidth > 1 should be an extension. Could we unify the same part and setup the beam search related parameters only when beamWidth > 1?

File: cpp/tests/unit_tests/layers/beamSearchLayerTest.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        auto const decodingDomain = tensorrt_llm::layers::DecoderDomain(
+            this->maxBatchSize(), params.beamWidth, this->mVocabSize, this->mVocabSizePadded);
+        this->mSamplingLayer
+            = std::make_shared<tensorrt_llm::layers::BeamSearchLayer<T>>(decodingDomain, this->mBufferManager);
+    }
+};
+
+TYPED_TEST_SUITE(BeamSearchLayerTest, FloatAndHalfTypes);
+
+TYPED_TEST(BeamSearchLayerTest, BeamWidth2)
```
Comment (byshiue) [inline cpp/tests/unit_tests/layers/beamSearchLayerTest.cpp:50]: Does it support variable beam width now? If yes, can you add one test for this feature?

File: cpp/tensorrt_llm/kernels/beamSearchKernels.h [hunk 9]
```diff
@@ -37,23 +39,33 @@ struct BeamHypotheses
...
     // Scalar values
-    bool bReturnNormedScore{false};         // return `normedScore` or `cumLogProbs`, always be `false` now
-    size_t nMaxBatchSize{0};                // buildtime max batch size
-    size_t nBatchSize{0};                   // runtime batch size
+    bool bReturnNormedScore{false};         // Return `normedScore` or `cumLogProbs`, always be `false` now
+    bool bVBWS{false};                      // whether to use VBWS for Beam-Search
+    size_t nMaxBatchSize{0};                // Buildtime max batch size
+    size_t nBatchSize{0};                   // Runtime batch size
     size_t nBeamWidth{0};                   //
+    size_t nBeamWidthIn{0};                 // Scalar value of current input beam width, for VBWS
```
Comment (byshiue) [inline cpp/tensorrt_llm/kernels/beamSearchKernels.h:50]: Why this is scalar instead of a vector?

File: cpp/tensorrt_llm/kernels/beamSearchKernels.cu [hunk 10]
```diff
@@ -85,71 +85,257 @@ template void invokeTopkBeamSearch<half, false>(
...
     //     column index in logProbs:            stage1Id[c][d]              -> 0    (e)
     //     pad for previous tokens:             b * nV                      -> 5    (f)
     //     final output:                        e + f                       -> 5
-
-    int const bid = blockIdx.x;
-    for (int j = threadIdx.x; j < nBM * 2; j += blockDim.x)
+    //
+    // ========================================================================================================
+    // Example for VBWS:
+    // nBS = 2, nBMIn = 3, nBMOut = 5, nBM = 7, nV = 11, use logProbs with integer values here for simplicity.
```
Comment (byshiue) [inline cpp/tensorrt_llm/kernels/beamSearchKernels.cu:198]: Why the shape of stage1LogProbs is [2, 3, 10]? Where the 10 come from? Is it from `2 * nBMOut`?

File: cpp/tensorrt_llm/layers/beamSearchLayer.cu (hunk 2 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/layers/beamSearchLayer.cu]: Redundant parenthesis are Removed.

File: cpp/tensorrt_llm/kernels/beamSearchKernels.cu (hunk 10 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/kernels/beamSearchKernels.cu:198]: The shape of stage1LogProbs is [2,3,11], which 11 is vocabSize of `logits`.

File: cpp/tensorrt_llm/kernels/beamSearchKernels.h (hunk 9 above)
Comment (wili-65535) [reply cpp/tensorrt_llm/kernels/beamSearchKernels.h:50]: Before Diverse-Beam-Width-Search (DBWS) is supported, the beam widths in a batch is the same, so we can use a scalar instead of vector temporarily.
After DBWS is supported, this variable will become a vector.

File: cpp/tests/unit_tests/layers/beamSearchLayerTest.cpp (hunk 8 above)
Comment (wili-65535) [reply cpp/tests/unit_tests/layers/beamSearchLayerTest.cpp:50]: Till this PR, VBWS is not able to run end-to-end.
We will add the unit test in the part4 of the whole PR.

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp (hunk 7 above)
Comment (wili-65535) [reply cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:241]: Sure, we can refactor the code of beam search related code in the next part of the PR.

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp (hunk 5 above)
Comment (wili-65535) [reply cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:216]: Sure, I will simplify the beam search related if-branch in the next part of the PR.

**Review by byshiue — APPROVED** [review]

File: cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp (hunk 6 above)
Comment (wili-65535) [reply cpp/tests/unit_tests/layers/baseSamplingLayerTest.cpp:269]: This is expected behavior, `mIdsPtrHost` is used for `decodeOutputs->outputIdsPtrHost`, which is different from sampling workflow.
